commons/src/main/java/io/aiven/kafka/connect/common/source/AbstractSourceTask.java

```diff
@@ -0,0 +1,510 @@
+        if (sourceRecordIterator.hasNext()) {
+            backoff.reset();
+            final SourceRecord sourceRecord = sourceRecordIterator.next();
+            if (logger.isDebugEnabled()) {
+                logger.debug("tryAdd() : read record " + sourceRecord.sourceOffset());
+            }
+            results.add(sourceRecord);
+            return true;
+        }
+        logger.info("No records found in tryAdd call");
+        return false;
+    }
+
+    /**
+     * Returns {@code true} if the connector is not stopped and the timer has not expired.
+     *
+     * @return {@code true} if the connector is not stopped and the timer has not expired.
+     */
+    protected boolean stillPolling() {
+        final boolean result = !connectorStopped.get() && !timer.isExpired();
+        logger.debug("Still polling: {}", result);
+        return result;
+    }
+
+    @Override
+    public final List<SourceRecord> poll() {
+        logger.debug("Polling");
+        if (connectorStopped.get()) {
+            logger.info("Stopping");
+            closeResources();
+            return Collections.emptyList();
+        } else {
+            timer.start();
+            try {
+                final List<SourceRecord> result = populateList();
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Poll() returning {} SourceRecords.", result == null ? null : result.size());
+                }
+                return result;
+            } finally {
+                timer.stop();
+                timer.reset();
+            }
+        }
+    }
+
+    /**
+     * Attempts to populate the return list. Will read as many records into the list as it can until the timer expires
+     * or the task is shut down.
+     *
+     * @return A list SourceRecords or {@code null} if the system hit a runtime exception.
+     */
+    private List<SourceRecord> populateList() {
+        final List<SourceRecord> results = new ArrayList<>();
+        try {
+            while (stillPolling() && results.size() < maxPollRecords) {
+                if (!tryAdd(results, sourceRecordIterator)) {
+                    if (!results.isEmpty()) {
+                        logger.debug("tryAdd() did not add to the list, returning current results.");
+                        // if we could not get a record and the results are not empty return them
+                        break;
+                    }
+                    logger.debug("Attempting {}", backoff);
+                    backoff.cleanDelay();
+                }
+            }
+
+        } catch (RuntimeException e) { // NOPMD must catch runtime here.
+            logger.error("Error during poll(): {}", e.getMessage(), e);
+            if (config.getErrorsTolerance() == ErrorsTolerance.NONE) {
+                logger.error("Stopping Task");
+                return null; // NOPMD must return null in this case.
```

Comment:
Based on this 
"Poll this source task for new records. If no data is currently available, this method should block but return control to the caller regularly (by returning null) in order for the task to transition to the PAUSED state if requested to do so.
The task will be stopped on a separate thread, and when that happens this method is expected to unblock, quickly finish up any remaining processing, and return.
Returns:
a list of source records"

returning null will not stop the task.